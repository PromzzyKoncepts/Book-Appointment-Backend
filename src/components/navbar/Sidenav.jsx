/* eslint-disable import/no-extraneous-dependencies */

import { useState } from 'react';
import { NavLink, useNavigate } from 'react-router-dom';
import Proptypes from 'prop-types';
import KeyboardDoubleArrowRightIcon from '@mui/icons-material/KeyboardDoubleArrowRight';
import KeyboardDoubleArrowLeftIcon from '@mui/icons-material/KeyboardDoubleArrowLeft';
import LogoutIcon from '@mui/icons-material/Logout';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import navData from './navData';
import Regal from '../../assets/regal.png';
import styles from './sidenav.module.css';

export default function Sidenav({ isLoggedIn, setIsLoggedIn }) {
  const [open, setopen] = useState(true);
  const toggleOpen = () => {
    setopen(!open);
  };
  
  const currentUserData = JSON.parse(localStorage.getItem('user'));
  const currentUser = currentUserData?.user;
  console.log(currentUser, "this is the current user")
  const displayCurrUser = {
    id: 0,
    icon: <AccountCircleIcon />,
    text: currentUser.name,
    link: '/',
    isProtected: true,
  }

  const logout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    setIsLoggedIn(false);
  };

  const navigateToLogin = () => {
    const navigate = useNavigate();
    navigate('/login');
  };

  const loginTab = {
    id: 6,
    icon: <LogoutIcon />,
    text: isLoggedIn ? 'Logout' : 'Login',
    link: isLoggedIn ? '/' : '/login',
    onClick: isLoggedIn ? () => logout() : () => navigateToLogin(),
    isProtected: false,
  };

  const tabsToDisplay = [displayCurrUser, ...navData, loginTab].filter((tab) => {
    if (isLoggedIn) return tab;
    return !tab.isProtected;
  });

  return (
    <div className={open ? styles.sidenav : styles.sidenavClosed}>
      <img src={Regal} alt="logo" className={open ? styles.logo : styles.logoClosed} />
      <button type="button" className={styles.menuBtn} onClick={toggleOpen}>
        {open ? <KeyboardDoubleArrowLeftIcon /> : <KeyboardDoubleArrowRightIcon />}
      </button>
      {tabsToDisplay.map((item) => (
        <NavLink key={item.id} className={styles.sideitem} to={item.link} onClick={item.onClick}>
          {item.icon}
          <span className={styles.linkText}>{item.text}</span>
        </NavLink>
      ))}
    </div>
  );
}

Sidenav.propTypes = {
  isLoggedIn: Proptypes.bool.isRequired,
  setIsLoggedIn: Proptypes.func.isRequired,
};
