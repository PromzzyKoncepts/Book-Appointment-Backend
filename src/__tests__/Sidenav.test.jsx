import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import store from '../redux/configureStore';
import '@testing-library/jest-dom/';
import Sidenav from '../components/navbar/Sidenav';

describe('SideNav component', () => {
  test('Should render the proper SideNav component', () => {
    const sideNavcomponent = render(
      <React.StrictMode>
        <Provider store={store}>
          <Router>
            <Sidenav />
          </Router>
        </Provider>
        ,
      </React.StrictMode>,
    );
    expect(sideNavcomponent).toMatchSnapshot();
  });
});
