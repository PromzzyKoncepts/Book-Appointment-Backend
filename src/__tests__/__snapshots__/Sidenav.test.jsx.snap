// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav component Should render the proper SideNav component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sidenav"
      >
        <img
          alt="logo"
          class="logo"
          src="regal.png"
        />
        <button
          class="menuBtn"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardDoubleArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17.59 18 19 16.59 14.42 12 19 7.41 17.59 6l-6 6z"
            />
            <path
              d="m11 18 1.41-1.41L7.83 12l4.58-4.59L11 6l-6 6z"
            />
          </svg>
        </button>
        <a
          aria-current="page"
          class="sideitem active"
          href="/"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="HomeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          <span
            class="linkText"
          >
            Home
          </span>
        </a>
        <a
          class="sideitem"
          href="/login"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LogoutIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
          <span
            class="linkText"
          >
            Login
          </span>
        </a>
      </div>
      ,
    </div>
  </body>,
  "container": <div>
    <div
      class="sidenav"
    >
      <img
        alt="logo"
        class="logo"
        src="regal.png"
      />
      <button
        class="menuBtn"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardDoubleArrowLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.59 18 19 16.59 14.42 12 19 7.41 17.59 6l-6 6z"
          />
          <path
            d="m11 18 1.41-1.41L7.83 12l4.58-4.59L11 6l-6 6z"
          />
        </svg>
      </button>
      <a
        aria-current="page"
        class="sideitem active"
        href="/"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
        <span
          class="linkText"
        >
          Home
        </span>
      </a>
      <a
        class="sideitem"
        href="/login"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LogoutIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
          />
        </svg>
        <span
          class="linkText"
        >
          Login
        </span>
      </a>
    </div>
    ,
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
